compiler/src/dotty/tools/dotc/core/Phases.scala

```diff
@@ -105,6 +105,10 @@ object Phases {
               prevPhases += phase.phaseName
               phase
             }
+
+          // synchronize symbol.defTree if necessary
+          if phaseToAdd.synchronizeDefTree then
+            fusedPhases += new SyncDefTree
```

Comment:
Improvements possible here: if phase `A` and `B` both require `SyncDefTree`, but there are no other phases between `A` and `B`, then we only need to insert one `SyncDefTree` phase before `A`. This avoids inserting a useless `SyncDefTree` before phase `B`.

One way to do that is to introduce another flag `val isCheckerPhase: Boolean`.

Reply:
Is the meaning of `isCheckerPhase` that a phase does not modify trees?

If so, should we then go further and enforce this with a new subclass of `Phase` whose `run` method does not update `ctx.compilationUnit.tpdTree`?